### ISSUE TITLE
[clang][bytecode] The copy a range-for makes is not destroyed, so its allocation is reported as leaked

### ISSUE THREAD

**Issue by j4niwzis** [llvm/llvm-project#221401]
Clang 22.1.8 (Debian, `22.1.8~++20260714014842+ca7933e47d3a`), target `x86_64-pc-linux-gnu`.

With `-fexperimental-new-constant-interpreter`, a range-based `for` whose loop variable is a
copy leaves that copy's allocation behind: the evaluation ends with "allocation performed here
was not deallocated". The default evaluator accepts the same code.

### Reproducer

```cpp
#include <vector>

struct holder {
  std::vector<int> values;
};

consteval int copies_in_a_loop() {
  std::vector<holder> source;
  source.push_back(holder{std::vector<int>{1}});
  source.push_back(holder{std::vector<int>{2}});
  int total = 0;
  // The loop variable is a copy. Each one is destroyed at the end of the
  // iteration, and its vector's allocation should go with it.
  for (holder item : source) total += static_cast<int>(item.values.size());
  return total;
}

static_assert(copies_in_a_loop() == 2);
int main() {}
```

```sh
clang++ -std=c++23 -fsyntax-only repro.cpp                                        # accepted
clang++ -std=c++23 -fsyntax-only -fexperimental-new-constant-interpreter repro.cpp
```

The second reports:

```
repro.cpp:18:15: error: static assertion expression is not an integral constant expression
   18 | static_assert(copies_in_a_loop() == 2);
      |               ^~~~~~~~~~~~~~~~~~~~~~~
__memory/allocator.h:99:13: note: allocation performed here was not deallocated
   99 |     return {allocate(__n), __n};
      |             ^~~~~~~~~~~~~
1 error generated.
```

### What narrows it down

All three of these are accepted by the bytecode interpreter, and differ from the failing case
only in how the copy is made or whether one is made at all:

```cpp
{ holder item = source[0]; total += static_cast<int>(item.values.size()); }         // copy in a block

for (std::size_t i = 0; i < source.size(); ++i) {                                   // copy in an index loop
  holder item = source[i];
  total += static_cast<int>(item.values.size());
}

for (const holder& item : source) total += static_cast<int>(item.values.size());    // reference in a range-for
```

So it is the range-for's own copy that is not destroyed. It also matters that the element owns
an allocation: with a `holder` of scalars there is nothing to leak and nothing is reported.

### Where it was met

A loop of the shape

```cpp
for (transition item : state.transitions) {
  ...
  if (...) { kept.push_back(std::move(item)); continue; }
  ...
}
```

which is the same thing with a move out of some iterations and not others -- the iterations
that do not move the copy away are the ones whose allocation is reported.
